src/main/java/at/hannibal2/skyhanni/features/skillprogress/SkillProgress.kt

```diff
@@ -0,0 +1,459 @@
+    @SubscribeEvent
+    fun onProfileSwitch(event: PreProfileSwitchEvent) {
+        display = emptyList()
+        allDisplay = emptyList()
+        etaDisplay = emptyList()
+        skillExpPercentage = 0.0
+    }
+
+    @SubscribeEvent
+    fun onTick(event: LorenzTickEvent) {
+        if (!isEnabled()) return
+        if (lastUpdate.passedSince() > 3.seconds) showDisplay = config.alwaysShow.get()
+
+        if (event.repeatSeconds(1)) {
+            allDisplay = formatAllDisplay(drawAllDisplay())
+            etaDisplay = drawETADisplay()
+        }
+
+        if (event.repeatSeconds(2)) {
+            update()
+            updateSkillInfo(activeSkill)
+        }
+    }
+
+    @SubscribeEvent
+    fun onLevelUp(event: SkillOverflowLevelupEvent) {
+        if (!isEnabled()) return
+        if (!config.overflowConfig.enableInChat) return
+        val skillName = event.skill.displayName
+        val oldLevel = event.oldLevel
+        val newLevel = event.newLevel
+        val skill = SkillAPI.storage?.get(event.skill) ?: return
+        val goalReached = newLevel == skill.customGoalLevel && customGoalConfig.enableInChat
+
+        val rewards = buildList {
+            add("  §r§7§8+§b1 Flexing Point")
+            if (newLevel % 5 == 0)
+                add("  §r§7§8+§d50 SkyHanni User Luck")
+        }
+        val messages = listOf(
+            "§3§l▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬",
+            "  §r§b§lSKILL LEVEL UP §3$skillName §8$oldLevel➜§3$newLevel",
+            if (goalReached)
+                listOf(
+                    "",
+                    "  §r§d§lGOAL REACHED!",
+                    "",
+                ).joinToString("\n") else
+                "",
+            "  §r§a§lREWARDS",
+            rewards.joinToString("\n"),
+            "§3§l▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬"
+        )
+
+        chat(messages.joinToString("\n"), false)
+
+        if (goalReached)
+            chat("§lYou have reached your goal level of §b§l${skill.customGoalLevel} §e§lin the §b§l$skillName §e§lskill!")
+
+        SoundUtils.createSound("random.levelup", 1f, 1f).playSound()
+    }
+
+    @SubscribeEvent
+    fun onConfigLoad(event: ConfigLoadEvent) {
+        onToggle(
+            config.enabled,
+            config.alwaysShow,
+            config.showActionLeft,
+            config.useIcon,
+            config.usePercentage,
+            config.useSkillName,
+            config.overflowConfig.enableInDisplay,
+            config.overflowConfig.enableInProgressBar,
+            config.overflowConfig.enableInEtaDisplay,
+            barConfig.enabled,
+            barConfig.useChroma,
+            barConfig.useTexturedBar,
+            allSkillConfig.enabled,
+            etaConfig.enabled
+        ) {
+            updateDisplay()
+            update()
+        }
+    }
+
+    @SubscribeEvent(priority = EventPriority.LOW)
+    fun onActionBar(event: ClientChatReceivedEvent) {
```

Comment:
Why are we using ClientChatReceivedEvent and not LorenzChatEvent here? is the cancel priority different to another mod? does the replace logic not working as expeted? if we replace with skyhannis tool, we see the replace cleanly in /shchathistory.

Reply:
LorenzChatEvent do not contain ActionBar messages.

I'm talking with Cal since he made the ActionBarUpdateEvent